crates/glaredb/src/lib.rs

```diff
@@ -0,0 +1,396 @@
+        Self {
+            location: None,
+            connection_target: None,
+            ..Default::default()
+        }
+    }
+
+    pub fn with_env_reader(&mut self, reader: Arc<Box<dyn EnvironmentReader>>) -> &mut Self {
+        self.environment_reader = Some(reader);
+        self
+    }
+
+    pub async fn connect(&self) -> Result<Connection, ExecError> {
+        let mut engine = Engine::from_backend(self.backend()).await?;
+
+        engine = engine.with_spill_path(self.spill_path.clone().map(|p| p.into()))?;
+
+        let mut session = engine.default_local_session_context().await?;
+
+        session
+            .create_client_session(
+                self.cloud_url(),
+                self.cloud_addr.clone().unwrap_or_default(),
+                self.disable_tls.unwrap_or_default(),
+                self.client_type.clone().unwrap_or_default(),
+                None,
+            )
+            .await?;
+        session.register_env_reader(self.environment_reader.clone());
+
+        Ok(Connection {
+            session: Arc::new(Mutex::new(session)),
+            _engine: Arc::new(engine),
+        })
+    }
+
+    fn backend(&self) -> EngineBackend {
+        if let Some(location) = self.location.clone() {
+            EngineBackend::Remote {
+                location,
+                options: self.storage_options.clone(),
+            }
+        } else if let Some(data_dir) = self.data_dir() {
+            EngineBackend::Local(data_dir)
+        } else {
+            EngineBackend::Memory
+        }
+    }
+
+    fn data_dir(&self) -> Option<PathBuf> {
+        match self.connection_target.clone() {
+            Some(s) => match Url::parse(&s) {
+                Ok(_) => None,
+                Err(_) => Some(PathBuf::from(s)),
+            },
+            None => None,
+        }
+    }
+
+    fn cloud_url(&self) -> Option<Url> {
+        self.connection_target
+            .clone()
+            .and_then(|v| Url::parse(&v).ok())
+    }
+}
+
+#[derive(Clone)]
+pub struct Connection {
+    session: Arc<Mutex<TrackedSession>>,
+    _engine: Arc<Engine>,
+}
+
+impl Connection {
+    pub fn execute(&self, query: impl Into<String>) -> Operation {
+        Operation {
+            op: OperationType::Execute,
+            query: query.into(),
+            conn: Arc::new(self.clone()),
+            schema: None,
+        }
+    }
+
+    pub fn sql(&self, query: impl Into<String>) -> Operation {
+        Operation {
+            op: OperationType::Sql,
+            query: query.into(),
+            conn: Arc::new(self.clone()),
+            schema: None,
+        }
+    }
+
+    pub fn prql(&self, query: impl Into<String>) -> Operation {
+        Operation {
+            op: OperationType::Prql,
+            query: query.into(),
+            conn: Arc::new(self.clone()),
+            schema: None,
+        }
+    }
+}
+
+pub struct RecordStream(Pin<Box<dyn Stream<Item = Result<RecordBatch, DataFusionError>> + Send>>);
```

Comment:
I would have this wrap a `SendableRecordBatchStream` since that's really just an alias for this. The `From` impl below will result in this the stream being double boxed, which probably isn't what we want.